pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -331,66 +329,48 @@ func (am *Alertmanager) buildIntegrationsMap(receivers []*apimodels.PostableApiR
+	receiverCfg, err := alertingNotify.BuildReceiverConfiguration(context.Background(), receiver, am.decryptFn)
+	if err != nil {
+		return nil, err
+	}
+	s := &sender{am.NotificationService}
+	img := newImageStore(am.Store)
+	integrations, err := alertingNotify.BuildReceiverIntegrations(
+		receiverCfg,
+		tmpl,
+		img,
+		LoggerFactory,
+		func(n receivers.Metadata) (receivers.WebhookSender, error) {
+			return s, nil
+		},
+		func(n receivers.Metadata) (receivers.EmailSender, error) {
+			return s, nil
+		},
+		am.orgID,
+		setting.BuildVersion,
+	)
+	if err != nil {
+		return nil, err
 	}
 	return integrations, nil
 }
 
-func (am *Alertmanager) buildReceiverIntegration(r *alertingNotify.GrafanaReceiver, tmpl *alertingNotify.Template) (alertingNotify.NotificationChannel, error) {
-	// secure settings are already encrypted at this point
-	secureSettings := make(map[string][]byte, len(r.SecureSettings))
-
-	for k, v := range r.SecureSettings {
-		d, err := base64.StdEncoding.DecodeString(v)
-		if err != nil {
-			return nil, alertingNotify.InvalidReceiverError{
-				Receiver: r,
-				Err:      errors.New("failed to decode secure setting"),
-			}
-		}
-		secureSettings[k] = d
+func (am *Alertmanager) buildReceiverIntegration(r *alertingNotify.GrafanaIntegrationConfig, tmpl *alertingTemplates.Template) (*alertingNotify.Integration, error) {
+	apiReceiver := &alertingNotify.APIReceiver{
+		GrafanaIntegrations: alertingNotify.GrafanaIntegrations{
+			Integrations: []*alertingNotify.GrafanaIntegrationConfig{r},
+		},
 	}
-
-	var (
-		cfg = &receivers.NotificationChannelConfig{
-			UID:                   r.UID,
-			OrgID:                 am.orgID,
-			Name:                  r.Name,
-			Type:                  r.Type,
-			DisableResolveMessage: r.DisableResolveMessage,
-			Settings:              r.Settings,
-			SecureSettings:        secureSettings,
-		}
-	)
-	factoryConfig, err := receivers.NewFactoryConfig(cfg, NewNotificationSender(am.NotificationService), am.decryptFn, tmpl, newImageStore(am.Store), LoggerFactory, setting.BuildVersion)
+	integrations, err := am.buildReceiverIntegrations(apiReceiver, tmpl)
 	if err != nil {
-		return nil, alertingNotify.InvalidReceiverError{
-			Receiver: r,
-			Err:      err,
-		}
-	}
-	receiverFactory, exists := alertingNotify.Factory(r.Type)
-	if !exists {
-		return nil, alertingNotify.InvalidReceiverError{
-			Receiver: r,
-			Err:      fmt.Errorf("notifier %s is not supported", r.Type),
-		}
+		return nil, err
 	}
-	n, err := receiverFactory(factoryConfig)
-	if err != nil {
-		return nil, alertingNotify.InvalidReceiverError{
-			Receiver: r,
-			Err:      err,
-		}
+	if len(integrations) == 0 {
+		// this should not happen but it is better return some error than panic
+		return nil, fmt.Errorf("failed to build integration")
 	}
-	return n, nil
+	return integrations[0], nil
```

Comment:
If I understand correctly, we have a slice of integrations, but in reality, it only has one element corresponding to the integration we want to test, right?
If so, a comment explaining this would really help. It's not so obvious why we're choosing only the first element of this slice.

Reply:
correct, this is related to my previous comment, this piece is also hacky and problematic, the only place where we use it is in testing API, which pieces were moved to alerting repository without any refactoring. Although the testing API accepts receiver configuration, the request handler code splits receiver into integrations and call the function for each integration. This function creates a 1-integration receiver, builds it, and returns the first one from the result. If we properly refactor testing API, we will be able to remove this confusing method